gluten-substrait/src/main/scala/org/apache/gluten/extension/GlutenPlan.scala

```diff
@@ -26,6 +26,7 @@ import org.apache.spark.sql.catalyst.trees.TreeNode
 sealed trait ValidationResult {
   def ok(): Boolean
   def reason(): String
+  def merge(other: ValidationResult): ValidationResult
```

Comment:
I see fallback tags already support appending: https://github.com/apache/incubator-gluten/blob/453cbebc26f5dfef76669f8abcda578820420352/gluten-core/src/main/scala/org/apache/gluten/extension/columnar/FallbackTag.scala#L37C14-L37C24

Can we do merge either only on tags or only on validation results? Say, either to make tags mergeable and then convert the merged tag to a big validation result, or to make tags non-mergeable then convert the tags to validation results that can be merged afterwards.

Reply:
If we want to consolidate the two, the validation failure should match the tagged operator. However, the validation result failures are not strictly bound to the tagged operator. For example in ShuffleExchangeExec, we add a sort before exchange for round-robin partitioning and the validation failure of the Sort will add a fallback tag on ColumnarShuffleExchange.

Merging the validation reason is more for the debugging purpose to give the developer more information by not returning the validation result immediately. As it's merged in a single validation call, it should be considered as one source of fallback. It seems that appending FallbackTags is intended to add multiple fallback reasons to an operator when fallback is triggered by different sources.